src/mlpack/core/data/load_numeric_csv.hpp

```diff
@@ -0,0 +1,154 @@
+ */
+#ifndef MLPACK_CORE_DATA_LOAD_NUMERIC_CSV_HPP
+#define MLPACK_CORE_DATA_LOAD_NUMERIC_CSV_HPP
+
+#include "load_csv.hpp"
+
+namespace mlpack{
+namespace data{
+
+template<typename eT>
+bool LoadCSV::ConvertToken(eT& val,
+                           const std::string& token)
+{
+  const size_t N = size_t(token.length());
+  // Fill empty data points with 0
+  if (N == 0)
+  {
+    val = eT(0);
+    return true;
+  }
+
+  const char* str = token.c_str();
+
+  // Checks for +/-INF and NAN
+  // Converts them to their equivalent representation
+  // from numeric_limits 
+  if ((N == 3) || (N == 4))
+  {
+    const bool neg = (str[0] == '-');
+    const bool pos = (str[0] == '+');
+
+    const size_t offset = ((neg || pos) && (N == 4)) ? 1 : 0;
+
+    const char sig_a = str[offset];
+    const char sig_b = str[offset+1];
+    const char sig_c = str[offset+2];
+
+    if (((sig_a == 'i') || (sig_a == 'I')) &&
+        ((sig_b == 'n') || (sig_b == 'N')) &&
+        ((sig_c == 'f') || (sig_c == 'F')))
+    {
+      val = neg ? -(std::numeric_limits<eT>
+                    ::infinity()) : std::numeric_limits<eT>::infinity();
+      return true;
+    }
+    else if (((sig_a == 'n') || (sig_a == 'N')) &&
+             ((sig_b == 'a') || (sig_b == 'A')) &&
+             ((sig_c == 'n') || (sig_c == 'N')))
+    {
+      val = std::numeric_limits<eT>::quiet_NaN();
+      return true;
+    }
+  }
+
+  char* endptr = nullptr;
+
+  // Convert the token into ccorrect type.
+  // If we have a eT as unsigned int,
+  // it will convert all negative numbers to 0
+  if (std::is_floating_point<eT>::value)
+  {
+    val = eT(std::strtod(str, &endptr));
+  }
+  else if (std::is_integral<eT>::value)
+  {
+    if (std::is_signed<eT>::value)
+      val = eT(std::strtoll(str, &endptr, 10));
+    else
+    {
+      if (str[0] == '-')
+      {
+        val = eT(0);
+        return true;
+      }
+      val = eT(std::strtoull(str, &endptr, 10));
+    }
+  }
+
+  if (str == endptr)
+    return false;
+
+  return true;
+}
+
+template<typename eT>
+bool LoadCSV::LoadNumericCSV(arma::Mat<eT>& x, std::fstream& f)
+{
+  bool load_okay = f.good();
+  f.clear();
+  std::pair<size_t, size_t> mat_size = GetMatrixSize<true>(f);
+  x.zeros(mat_size.first, mat_size.second);
+  size_t row = 0;
+
+  std::string lineString;
+  std::stringstream lineStream;
+  std::string token;
+
+  while (f.good())
+  {
+    // Parse the file line by line
+    std::getline(f, lineString);
+
+    if (lineString.size() == 0)
+      break;
+
+    lineStream.clear();
+    lineStream.str(lineString);
+
+    size_t col = 0;
+
+    while (lineStream.good())
+    {
+      // Parse each line
+      std::getline(lineStream, token, ',');
+
+      // This will handle loading of both dense and sparse.
+      // Initialize tmp_val of type eT with value 0.
+      eT tmp_val = eT(0);
+
+      if (ConvertToken<eT>(tmp_val, token))
+      {
+        x.at(row, col) = tmp_val;
+        ++col;
+      }
```

Comment:
@rcurtin I don't think this can ever fail, i.e. return `false`.

[This](https://github.com/mlpack/mlpack/blob/933b68c2b30ca2bed06f1c6b41597674e117757e/src/mlpack/core/data/load_numeric_csv.hpp#L89) is the only condition that will fail `convert_token` and will make it return false.

But I don't think `str` can ever be equal to `nullptr`.
`str` is [`const char* str = token.c_str()`](https://github.com/mlpack/mlpack/blob/933b68c2b30ca2bed06f1c6b41597674e117757e/src/mlpack/core/data/load_numeric_csv.hpp#L32) and I don't think `string::c_str()` ever returns a `nullptr`.

For [reference](https://stackoverflow.com/questions/45269038/could-stdstringc-str-ever-return-nullptr). I might be wrong let me know what you think.

Reply:
Should we return `false` when neither of the following code blocks succeed though?

```c++
  // Convert the token into ccorrect type.
  // If we have a eT as unsigned int,
  // it will convert all negative numbers to 0
  if (std::is_floating_point<eT>::value)
  {
    val = eT(std::strtod(str, &endptr));
  }
  else if (std::is_integral<eT>::value)
  {
    if (std::is_signed<eT>::value)
      val = eT(std::strtoll(str, &endptr, 10));
    else
    {
      if (str[0] == '-')
      {
        val = eT(0);
        return true;
      }
      val = eT(std::strtoull(str, &endptr, 10));
    }
  }
```

What I mean is, should we have an `else` like this?

```
else
{
  // Conversion failed, so return false.
  return false;
}
```

Reply:
I think we can do that, so [this](https://github.com/mlpack/mlpack/blob/933b68c2b30ca2bed06f1c6b41597674e117757e/src/mlpack/core/data/load_numeric_csv.hpp#L92) line will get executed only when the compiler gets to [this](https://github.com/mlpack/mlpack/blob/933b68c2b30ca2bed06f1c6b41597674e117757e/src/mlpack/core/data/load_numeric_csv.hpp#L85).

So I can move that return true inside the else condition and then add another else.
I will give it a try and push the code.

Reply:
@heisenbuug I agree with @rcurtin on this point, if you are sure that the conversion will never fail, then there is no need to put the `else  {return false}`.

Reply:
So I think there are two issues here we should address---first, it's probably useful to have an `else` block like I pointed out in case neither `std::is_floating_point<eT>::value` nor `std::is_integral<eT>::value` are `true` (in this case, an invalid `eT` has been specified).

Second, as you pointed out, we definitely *can* have situations where `convert_token()` returns `false` because `strtod()` or `strtoll()` failed.  In these situations, we should probably have some error handling here, in `LoadNumericCSV()`.  In this case, we probably want to set `load_okay` to `false`, since the load did not happen successfully---there was a conversion failure.  We could also issue a message to `Log::Warn` indicating the failed token (and if possible, its position).

Do you think we can fix that before merge?  Otherwise users might end up with their CSVs loading and appearing to succeed, but silently failing and putting 0s in place of invalid tokens (which is incorrect). :+1:

Reply:
@heisenbuug would you handle these cases please, once this one is handled, I will merge this pull request

Reply:
@shrit
* I already added a `false` condition at the end.
```
if (std::is_floating_point<eT>::value)
{...}
else if (std::is_integral<eT>::value)
{...}
// If none of the above conditions was executed,
// then the conversion will fail.
else
  return false;
```
* @rcurtin the point you mentioned regarding `strtod()` or `strtoll()` is already handled. [`This`](https://github.com/mlpack/mlpack/blob/933b68c2b30ca2bed06f1c6b41597674e117757e/src/mlpack/core/data/load_numeric_csv.hpp#L89) line handles it.
When conversion fails, `strtoll` `strtod` sets `endpt == str`.

Reply:
@heisenbuug thanks for that clarification!  I had overlooked that.  Do you think you could add a comment indicating that that `endptr == str` check is to catch failed `strtod()` or `strtoll()` calls?

There is only one more thing---the block of code that calls `ConvertToken()` looks like this:

```c++
    while (lineStream.good())
    {
      // Parse each line.
      std::getline(lineStream, token, ',');

      // This will handle loading of both dense and sparse.
      // Initialize tmp_val of type eT with value 0.
      eT tmp_val = eT(0);

      if (ConvertToken<eT>(tmp_val, token))
      {
        x.at(row, col) = tmp_val;
        ++col;
      }
    }
```

But, if `ConvertToken()` fails, we aren't setting `load_okay` to `false` or printing a warning or anything like that.  So, even though we are properly catching when a conversion is failing in `ConvertToken()`, we might still report `load_okay` to be `true` incorrectly, even though conversion failed somewhere in the file.  I might suggest a block more like this:

```c++
    while (lineStream.good())
    {
      // Parse each line.
      std::getline(lineStream, token, ',');

      // This will handle loading of both dense and sparse.
      // Initialize tmp_val of type eT with value 0.
      eT tmp_val = eT(0);

      if (ConvertToken<eT>(tmp_val, token))
      {
        x.at(row, col) = tmp_val;
        ++col;
      }
      else
      {
        // This error message could definitely be improved!  Ideally we could even print the token that failed to convert,
        // plus the original position in the CSV file so the user can easily find and fix the file.
        Log::Warn << "Failed to load CSV at row " << row << ", column " << col << " of matrix!";
        return false; // Terminate early if we failed to parse anything; there's no need to keep going.
      }
    }
```

Anyway, that is just an idea, so maybe you can adapt it into something better.

I'll be gone for the next two weeks on vacation, so don't feel obligated to wait for another response from me.  Other than this issue plus minor style issues, I think everything in this PR looks great and I am very excited about it getting merged. :tada: